src/lib/projectConfigValidation.ts

```diff
@@ -0,0 +1,439 @@
+      nodePath.push(error.params.property);
+      onlyKeyNode = true;
+      break;
+    case AjvErrorKeyword.Enum:
+      // Enum errors often happen on types and have cascading effects. We let users know that other errors can lead to this one.
+      if (error.params.allowedValues) {
+        message = `Try one of: ${error.params.allowedValues.join(', ')}`;
+      } else {
+        message = `Value not allowed. Other errors can cause this`;
+      }
+      break;
+    case AjvErrorKeyword.Maximum:
+    case AjvErrorKeyword.Minimum:
+    case AjvErrorKeyword.ExclusiveMaximum:
+    case AjvErrorKeyword.ExclusiveMinimum:
+      message = `This value must ${error.params.comparison}${error.params.limit}`;
+      break;
+    case AjvErrorKeyword.MaxItems:
+      message = `Value is too long, needs at most ${error.params.limit} items`;
+      break;
+    case AjvErrorKeyword.MaxLength:
+      message = `Value is too long, needs at most ${error.params.limit} characters`;
+      break;
+    case AjvErrorKeyword.MaxProperties:
+      message = `Too many properties, need at most ${error.params.limit}`;
+      break;
+    case AjvErrorKeyword.MinItems:
+      message = `Value is too short, needs at least ${error.params.limit} items`;
+      break;
+    case AjvErrorKeyword.MinLength:
+      message = `Value is too short, needs at least ${error.params.limit} characters`;
+      break;
+    case AjvErrorKeyword.MinProperties:
+      message = `Too few properties, need at least ${error.params.limit}`;
+      break;
+    case AjvErrorKeyword.MultipleOf:
+      message = `This value must be a multiple of ${error.params.multipleOf}`;
+      break;
+    case AjvErrorKeyword.OneOf:
+      onlyKeyNode = true;
+      message = `Does not match any of the allowed schemas`;
+      break;
+    case AjvErrorKeyword.Pattern:
+      message = `Value does not match pattern: ${error.params.pattern}`;
+      break;
+    case AjvErrorKeyword.Required:
+      onlyKeyNode = true;
+      message = getMissingRequiredFieldError(error.params.missingProperty);
+      break;
+    case AjvErrorKeyword.Type:
+      message = `Incorrect type: ${error.message}`;
+      break;
+    default:
+      break;
+  }
+  return { message, nodePath, onlyKeyNode };
+}
+
+function diagnosticDeduplicationKey(
+  nodePath: (string | number)[],
+  keyword: string,
+  params: any
+): string {
+  return `${keyword}::${nodePath.join('/')}::${JSON.stringify(params)}`;
+}
+
+function getKeyNode(node: jsonc.Node): jsonc.Node | undefined {
+  return node.parent?.type === 'property' ? node.parent.children?.[0] : node;
+}
+
+async function getCachedSchemasForAllTypes(
+  projectRoot: Uri,
+  platformVersion: string,
+  accountId: number
+): Promise<SchemaCacheEntry> {
+  // If schema cache has a schema for this version, return it
+  // If there is an error, we're going to return that too
+  if (schemaCache[platformVersion]) {
+    return schemaCache[platformVersion];
+  }
```

Comment:
What is the lifetime of this cache?  Is it as long as the editor is open?  Maybe we should consider timing items out of the cache

Reply:
While the editor is open or until someone reloads the window. We tell them this in the error message if we fail to fetch schemas.

It would be an unpleasant editor experience to have the schema change behind your back. Imagine you didn't have an error but now you do mid-typing? Or you leave the buffer and it's in a good state, come back and there's an error because the schemas silently updated? And you did nothing to change anything? I think it would be better if schema updates happened at some obvious point in the lifecycle of the editor, like on reload or restart.